atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/paging/CellPagerBatchSizingStrategy.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra.paging;
+
+public class CellPagerBatchSizingStrategy {
+    public PageSizes computePageSizes(int cellBatchHint, StatsAccumulator stats) {
+        if (stats.count() == 0) {
+            // No data yet: go with a square
+            return negotiateExactPageSize(Math.round(Math.sqrt(cellBatchHint)), cellBatchHint);
+        } else {
+            return computePageSizesAssumingNormalDistribution(
+                    cellBatchHint, stats.mean(), stats.populationStandardDeviation());
+        }
+    }
+
+    /* Given the batch size in cells supplied by the user, we want to figure out the optimal (or at least decent)
```

Comment:
First, this is a Cassandra API deficiency: for multiget_slice, you can only specify a single slice predicate for all rows you are requesting (i.e. start with the same column for all rows) . If you have several partially fetched rows, each of them has a different column you have to restart with. So you can't move the "rectangle" horizontally, only vertically.

Second, even if you could do that, you probably still wouldn't because of memory concerns: we need to return results in order, so you can't return the second row until you returned the first one. So you would need to keep everything you fetched in memory until the first row is done, which will result in OOMs.